tensorboard/plugins/projector/projector_plugin.py

```diff
@@ -173,8 +173,10 @@ def _latest_checkpoints_changed(configs, run_path_pairs):
       config = configs[run_name]
 
     # See if you can find a checkpoint file in the logdir.
-    logdir = _assets_dir_to_logdir(assets_dir)
```

Comment:
I think this can be fixed just by not calling _assets_dir_to_logdir() only when `assets_dir == self.logdir` - because every assets_dir in run_paths_pairs is a "plugins" subdirectory except the one that's for the actual logdir.

So it should suffice to just do

```py
logdir = self.logdir if assets_dir == self.logdir else _assets_dir_to_logdir(assets_dir)
```

Ditto below.

Reply:
Dear Nich, 
         Thanks for your suggestion very much. I updated the code here. I also consider the relative/absolute path of logdir, and logdir+run=assets_dir including your sample code.
It is working well in  various kinds of test cases in my environment.
Would u review again kindly,plz?

Reply:
For _latest_checkpoints_changed, I only see a single callsite in the configs method, so passing in the logdir should work fine.  If you got a build or unittest failure, can you specify what it was in more detail or give me a way to replicate it?

For the other place, I don't think the call to os.path.abspath() is needed and the check `if logdir in assets_dir: logdir = assets_dir` looks incorrect to me because it would include cases where the logdir is a prefix of the assets_dir, which should be true in all cases, so it would break the case where the checkpoints are in the assets dir.

My understanding of the original issue you raised was that calling `logdir = _assets_dir_to_logdir(assets_dir)` is incorrect in the case where there are no asset dirs, so self.logdir gets used as the only asset dir, with the following code in the configs method:

```py
# If there are no summary event files, the projector should still work,
# treating the `logdir` as the model checkpoint directory.
if not run_path_pairs:
   run_path_pairs.append(('.', self.logdir))
```

In particular, calling _assets_dir_to_logdir() will do the wrong thing if self.logdir incidentally contains a path segment `/plugins/` even though it's not a plugins asset directory.

My proposal was to fix this by skipping _assets_dir_to_logdir() if the assets dir equals self.logdir exactly, since that's the precise case where it shouldn't be called.

If you have a different understanding of the issue or there's a reason this fix doesn't work for you, can you please elaborate further?

Reply:
I'm still not following how you're arriving at a log directory structure that produces the behavior you describe.  I just did some testing and I can't reproduce the original issue at all.  Can you provide some instructions to reproduce the exact issue you're seeing?

In detail: based on your explanation, what I'm hearing is that you are getting run_path_pairs not from the fallback logic that says `if not run_paths_pairs: run_path_pairs.append(('.', self.logdir))` (which was my initial interpetation) but instead that it's coming from `_append_plugin_asset_directories()`.

However, the definition of that helper contains the following lines:
```py
assets_dir = os.path.join(self.run_paths[run], _PLUGINS_DIR, _PLUGIN_NAME)
assets_path_pair = (run, os.path.abspath(assets_dir))
run_path_pairs.append(assets_path_pair)
```

This would mean that every path added to `run_path_pairs` via this method was already extended to include the trailing path `/plugins/org_tensorflow_tensorboard_projector`.  In which case the logic in `_assets_dir_to_logdir()` should correctly strip that path suffix.  However, your example of `run_path_pairs` doesn't show that.

So I don't understand how to produce the situation you're describing.

Reply:
hi,reproduce steps:
1. make embeddings and save ckpt in 
```
C:\Users\hzhang16\share\NLP\ai\plugins\logs\tensorboard\1\embeddings\model.ckpt-1875.meta
```
 and other files like *.index,*.pbtxt,checkpoint, etc.

2. ```cd dir: C:\Users\hzhang16\share\NLP\ai\plugins\logs\tensorboard\1```

3. ```execute command : tensorboard --logdir=embeddings --port=8888```

4. the code and log is as below:
```
   @property
    def configs(self):
        """Returns a map of run paths to `ProjectorConfig` protos."""
        run_path_pairs = list(self.run_paths.items())
        print("run_path_pairs before:",run_path_pairs)###########log here
        self._append_plugin_asset_directories(run_path_pairs)
        print("run_path_pairs after:",run_path_pairs)###########log here
        # If there are no summary event files, the projector should still work,
        # treating the `logdir` as the model checkpoint directory.
        if not run_path_pairs:
            run_path_pairs.append(('.', self.logdir))
        if (self._run_paths_changed() or
                _latest_checkpoints_changed(self._configs, run_path_pairs, self.logdir)):
            self.readers = {}
            print("self.logdir:",self.logdir)###########log here
            self._configs, self.config_fpaths = self._read_latest_config_files(
                run_path_pairs)
            self._augment_configs_with_checkpoint_info()
        return self._configs
```

log:
run_path_pairs before: [('.', 'C:\\Users\\hzhang16\\share\\NLP\\ai\\plugins\\logs\\tensorboard\\1\\embeddings')]
run_path_pairs after: [('.', 'C:\\Users\\hzhang16\\share\\NLP\\ai\\plugins\\logs\\tensorboard\\1\\embeddings')]
self.logdir: embeddings

```
def _append_plugin_asset_directories(self, run_path_pairs):
        for run, assets in self.multiplexer.PluginAssets(_PLUGIN_NAME).items():
            print('run:',run)###########log here
            print('assets:',assets)###########log here
            if PROJECTOR_FILENAME not in assets:
                continue
            assets_dir = os.path.join(self.run_paths[run], _PLUGINS_DIR, _PLUGIN_NAME)
            assets_path_pair = (run, os.path.abspath(assets_dir))
            run_path_pairs.append(assets_path_pair)`
```
log:
run: .
assets: []

```    def _read_latest_config_files(self, run_path_pairs):
        """Reads and returns the projector config files in every run directory."""
        configs = {}
        config_fpaths = {}
        for run_name, assets_dir in run_path_pairs:
            config = ProjectorConfig()
            config_fpath = os.path.join(assets_dir, PROJECTOR_FILENAME)
            if tf.gfile.Exists(config_fpath):
                with tf.gfile.GFile(config_fpath, 'r') as f:
                    file_content = f.read()
                text_format.Merge(file_content, config)
            has_tensor_files = False
            for embedding in config.embeddings:
                if embedding.tensor_path:
                    if not embedding.tensor_name:
                        embedding.tensor_name = os.path.basename(embedding.tensor_path)
                    has_tensor_files = True
                    break

            if not config.model_checkpoint_path:
                # See if you can find a checkpoint file in the logdir.
                logdir = self.logdir
                print("logdir:",logdir)###########log here
                print("assets_dir", assets_dir)###########log here
                if os.path.abspath(logdir) != assets_dir:
                    logdir = _assets_dir_to_logdir(assets_dir)
                ckpt_path = _find_latest_checkpoint(logdir)`
```

log:
logdir: embeddings
assets_dir C:\Users\hzhang16\share\NLP\ai\plugins\logs\tensorboard\1\embeddings

It is my test step and debugging logs.
Would you check it plz?
Thanks!

Reply:
@nfelt 
Would you check the code and log kindly, plz?
Thanks a lot for your reply.